File: backend/app/schemas/reports.py
"""Pydantic v2 schemas for /reports (Phase 9.1).

Generic report snapshot — one shape feeds JSON view + Excel/PDF/CSV export.
Money/date cells are pre-formatted strings (BE formats via _fmt_money/_fmt_date).
Spec: docs/15-page-otchety.md, docs/31 §6, docs/32 §8. Decisions: docs/superpowers/specs/2026-06-28-phase9-reports-design.md.
"""
from __future__ import annotations

from typing import Optional, Union

from pydantic import BaseModel, ConfigDict, Field


class CellObj(BaseModel):
    """Styled cell. `text` is the export/plain render; FE uses kind + extras."""
    text: Optional[str] = None
    kind: Optional[str] = None     # claim|mono|text|stage|days|money|date|date-late|percent|note
    color: Optional[str] = None    # for kind='stage': token like '--proc'
    level: Optional[str] = None    # for kind='days': ''|'warn'|'bad'
    code: Optional[str] = None     # for kind='claim': parent code 'Т-67'
    title: Optional[str] = None    # for kind='claim': title


# A cell is either a plain string or a styled object.
Cell = Union[str, CellObj]


class Column(BaseModel):
    key: str
    label: str
    kind: Optional[str] = None
    align: Optional[str] = None    # 'left' | 'right'


class Section(BaseModel):
    title: Optional[str] = None
    columns: list[Column]
    rows: list[list[Cell]]
    footer: Optional[list[Cell]] = None


class Kpi(BaseModel):
    label: str
    value: str
    color: Optional[str] = None


class PeriodInfo(BaseModel):
    model_config = ConfigDict(populate_by_name=True)
    key: str
    label: str
    from_: Optional[str] = Field(default=None, alias="from")
    to: Optional[str] = None


class ReportOut(BaseModel):
    type: str
    title: str
    period: Optional[PeriodInfo] = None
    kpis: list[Kpi]
    sections: list[Section]


class FiltersOut(BaseModel):
    mtr: list[str]
    supplier: list[str]
    author: list[str]


__all__ = [
    "CellObj", "Cell", "Column", "Section", "Kpi", "PeriodInfo",
    "ReportOut", "FiltersOut",
]

File: frontend/src/api/reports.ts
import { apiFetch } from './client'
import { buildQuery } from './support'

// Зеркало backend schemas/reports.py (Фаза 9.1). Слепок отчёта — generic-форма;
// money/date cells — предформатированные строки (BE).

export type ReportType = 'time' | 'sums' | 'late' | 'people'
export type ExportFormat = 'excel' | 'pdf' | 'csv'
export type ReportPeriodKey = 'month' | 'quarter' | 'year' | 'custom'

export type CellObj = {
  text?: string | null
  kind?: string | null      // claim|mono|text|stage|days|money|date|date-late|percent|note
  color?: string | null     // kind='stage': '--proc'
  level?: string | null     // kind='days': ''|'warn'|'bad'
  code?: string | null      // kind='claim': parent code
  title?: string | null     // kind='claim': title
}
export type Cell = string | CellObj

export type Column = { key: string; label: string; kind?: string | null; align?: string | null }
export type Section = {
  title?: string | null
  columns: Column[]
  rows: Cell[][]
  footer?: Cell[] | null
}
export type Kpi = { label: string; value: string; color?: string | null }
// BE serializes PeriodInfo.from_ with alias "from" → JSON key is "from".
export type PeriodInfo = { key: string; label: string; from?: string | null; to?: string | null }
export type ReportSnapshot = {
  type: string
  title: string
  period: PeriodInfo | null
  kpis: Kpi[]
  sections: Section[]
}
export type Filters = { mtr: string[]; supplier: string[]; author: string[] }
export type ReportFilters = {
  period?: ReportPeriodKey
  date_from?: string
  date_to?: string
  mtr?: string
  supplier?: string
  author?: string
}

export function getReport(type: ReportType, flt: ReportFilters): Promise<ReportSnapshot> {
  return apiFetch<ReportSnapshot>(`/reports/${type}${buildQuery(flt)}`)
}

export function getFilters(): Promise<Filters> {
  return apiFetch<Filters>('/reports')
}

// File download — bypasses apiFetch (which parses JSON); we need the binary blob.
export async function downloadReport(
  type: ReportType,
  flt: ReportFilters,
  format: ExportFormat,
): Promise<void> {
  const q = buildQuery({ ...flt, format })
  const res = await fetch(`/api/reports/${type}/export${q}`, { credentials: 'include' })
  if (!res.ok) {
    throw { status: res.status, body: await res.json().catch(() => null) }
  }
  const blob = await res.blob()
  const disp = res.headers.get('content-disposition') || ''
  const m = /filename="?([^"]+)"?/i.exec(disp)
  const filename =
    m?.[1] ||
    `otchety_${type}.${format === 'excel' ? 'xlsx' : format}`
  const url = URL.createObjectURL(blob)
  const a = document.createElement('a')
  a.href = url
  a.download = filename
  document.body.appendChild(a)
  a.click()
  a.remove()
  URL.revokeObjectURL(url)
}
